partiql-spi/src/main/java/org/partiql/spi/value/DatumReader.java

```diff
@@ -0,0 +1,21 @@
+package org.partiql.spi.value;
+
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * The {@link DatumReader} interface is a low-level reader interface for reading streams of PartiQL data.
+ * <br>
+ * {@see java.io.Reader}
+ * <br>
+ * TODO
+ *  - public void reset();
+ *  - public void skip(long n);
+ */
+public interface DatumReader extends AutoCloseable {
+
+    /**
+     * @return next Datum or null.
+     */
+    @Nullable
+    public Datum next();
```

Comment:
Why wouldn't this look a bit more like the `IonReader` with `stepIn()` and `stepOut()`?

Kind of similar to `IonReader`, the constructors for `IonTextReader` takes in an IonValue or stream of data , and it doesn't return `IonValues` -- it returns primitives as you walk the data. I assumed this would be similar. They do expose an `IonElementLoader` which seemingly wraps a reader and creates an `IonElement`, but it's different than the reader itself.

As a side, this came into my mind because I saw the JSON directory placeholder. JSON doesn't allow for multiple top-level values, so the `DatumReader` would always just return a single `Datum`. As a consumer of this, you'd still need to know the in's and out's of the Datum API.

Reply:
The reader/write are really the two functions "encode" and "decode" you could conceive of backing them by a cursor-like API, but that is not the intention here. Notice how the methods return datums and not primitives. These encode a Datum into some bytes, and can decode back into the Datum interface.

The consumer of a reader is not responsible for how the value is traversed, that's what the Datum interface allows for them. This is simple about getting bytes into and outof the Datum interface.

I don't believe you would need to know the in's and out's of Datum to use this. It just wraps a stream of one-or-more values. Yes, json is one top-level value, but there's others like

* jsonlines
* yaml
* csv
* etc.

which allow for multiple values in a stream